architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,329 @@
+| User-defined type    | A resource-type that can be defined or modified by end-users.                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
+| OpenAPI/Schema       | A format for documenting HTTP-based APIs. We use OpenAPI v2 currently in Radius for client-sdk generation and validation of inbound requests. In this document we're primarily concerned with the parts of OpenAPI pertaining to request/response bodies.                                                                                                                                                                                                                                                             |
+| Declarative          | Functionality based on static definition of schemas and static definition of behaviors. In contrast with imperative functionality. Example: validation of a request body based on an OpenAPI schema is declarative validation.                                                                                                                                                                                                                                                                                        |
+| Tracked resource     | A resource type that is *tracked* by UCP. Tracked resources are cached by UCP and their existence can easily be queried without sending a request to the resource provider. All existing resource types in Radius are tracked resources.                                                                                                                                                                                                                                                                              |
+| Normal resource type | A resource type that is part of the `radius` plane and exists inside a resource group. Also has the `TrackedResource` set of fields and behaviors. This is the case for most resource types in Radius today. Normal resource types are our scenario priority because they are optimized for being part of an application. The resource types that are not *normal* are not usually part of an application, for example `System.AWS/credentials` is not part of an application and does not exist in a resource group. |
+
+## Objectives
+
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6688
+
+### Guiding principles
```

Comment:
should we have a guiding principle around cultivating the ecosystem for the community to contribute with low overhead and reduced dependency from Radius team

Reply:
I think that's captured by this part:

> We want to build a big ecosystem and so user-defined types should be easy to adopt and easy to develop.

I don't really want to capture community-building activities as part of the technical design.